src/software/logger/proto_logger.cpp

```diff
@@ -0,0 +1,191 @@
+                                      const std::string& serialized_proto)
+{
+    buffer_.push({
+        .protobuf_type_full_name = protobuf_type_full_name,
+        .serialized_proto        = serialized_proto,
+        .receive_time_sec        = time_provider_() - start_time_,
+    });
+}
+
+void ProtoLogger::logProtobufs()
+{
+    unsigned int replay_index = 0;
+
+    while (!shouldStopLogging())
+    {
+        std::string log_file_path =
+            log_folder_ + std::to_string(replay_index) + "." + REPLAY_FILE_EXTENSION;
+
+        gzFile gz_file = gzopen(log_file_path.c_str(), "wb");
+        if (!gz_file)
+        {
+            std::cerr << "ProtoLogger: Failed to open gzip log file: " << log_file_path
+                      << " Error: " + std::string(strerror(errno))
+                      << "\nStopping ProtoLogger logger thread!" << std::endl;
+            return;
+        }
+
+        // Start every replay file with the metadata, which includes the file format
+        // version. This allows us to keep backwards compatibility as the replay file
+        // format evolves.
+        std::string file_metadata =
+            REPLAY_FILE_VERSION_PREFIX + std::to_string(REPLAY_FILE_VERSION) + "\n";
+        int num_bytes_written = gzwrite(gz_file, file_metadata.c_str(),
+                                        static_cast<unsigned>(file_metadata.size()));
+        if (num_bytes_written != static_cast<int>(file_metadata.size()))
+        {
+            std::cerr << "ProtoLogger: Failed to write metadata to log file: "
+                      << log_file_path << std::endl;
+        }
+
+        while (!shouldStopLogging())
+        {
+            auto serialized_proto_opt =
+                buffer_.popLeastRecentlyAddedValue(BUFFER_BLOCK_TIMEOUT);
+            if (!serialized_proto_opt.has_value())
+            {
+                // Timed out without getting a new value
+                continue;
+            }
+
+            const auto& [proto_full_name, serialized_proto, receive_time_sec] =
+                serialized_proto_opt.value();
+
+            // Write the log entry to the file with the format:
+            std::string log_entry =
+                createLogEntry(proto_full_name, serialized_proto, receive_time_sec);
+            num_bytes_written = gzwrite(gz_file, log_entry.c_str(),
+                                        static_cast<unsigned>(log_entry.size()));
+
+            // Check if write was successful
+            if (num_bytes_written != static_cast<int>(log_entry.size()))
+            {
+                if (num_failed_logs_++ % FAILED_LOG_PRINT_FREQUENCY == 0)
+                {
+                    std::cerr << "ProtoLogger: Failed to write " << proto_full_name
+                              << " to log file: " << log_file_path << " "
+                              << std::to_string(num_failed_logs_) << " times"
+                              << std::endl;
+                }
```

Comment:
```suggestion
                if (failed_log_frequency_counter == 0)
                {
                    std::cerr << "ProtoLogger: Failed to write " << proto_full_name
                              << " to log file: " << log_file_path << " "
                              << std::to_string(num_failed_logs_) << " times"
                              << std::endl;
                }
                failed_logs_frequency_counter = (failed_logs_frequency_counter+1) % FAILED_LOG_PRINT_FREQUENCY
```

the variable doesn't quite match its functions. Also the nuances of when the `++` counter increments and stores the variable hurts my brain

Reply:
Yeah inline ++ is not the most readable. Updated to code